R/transformAbundance.R

```diff
@@ -0,0 +1,400 @@
+    assay <- assay * scale
+
+    # Function is selected based on the "transform" variable
+    FUN <- switch(transform,
+                  relabundance = .get_relabundance_table,
+                  Z = .get_z_table,
+                  log10 = .get_log10_table,
+                  log10p = .get_log10p_table,
+                  hellinger = .get_hellinger_table,
+                  identity = .get_identity_table,
+                  clr = .get_clr_table)
+
+    # Does the function call, arguments are "assay" abundance table and "pseudocount"
+    do.call(FUN,
+            list(assay = assay,
+                 target = target))
+}
+
+.get_relabundance_table <- function(assay, target){
+
+    # Calculates the relative abundances
+    mat <- sweep(assay, 2, colSums(assay), "/")
+    return(mat)
+}
+
+.get_z_table <- function(assay, target){
+
+    # Log10 can not be calculated if there is zero
+    if (any(assay == 0)) {
+        stop("Abundance table contains zero and Z transformation
+                is being applied without pseudocount. Try Z with
+                pseudocount=TRUE.")
+    }
+
+    # Gets the log transformed table
+    mat <- .get_log10_table(assay, target)
+
+    if(target=="features"){
+        # Z transform for features. Centers the feature data. After that, divides with
+        # the standard deviation of feature.
+        trans <- t(scale(t(mat)))
+        # Saves all features that are undetectable in every sample i.e. are NA
+        undetectables <- which(rowMeans(is.na(trans)) == 1)
+
+        # If there are some deatures that are undetectable
+        # i.e. "undetectable" has length over 0, and table have zeros
+        if (length(undetectables) > 0 & min(mat) == 0) {
+
+            warning("Some features were not detectable. In all samples, signal was
+                    under detectable limit.")
+
+            # Some features are undetectable in all samples, they are NA when scaled.
+            # 0 is assigned to those features.
+
+            trans[names(undetectables), ] <- 0
+
+            mat <- trans
+
+            # Deletes extra information
+            attr(mat,"scaled:scale") <- NULL
+            attr(mat,"scaled:center") <- NULL
+        }
+
+    }else if(target=="samples"){
+        # Performs z transformation for samples
+        mat <- apply(mat, 2, function(x) {
+            (x - mean(x))/sd(x)
+        })
+    }
+
+    return(mat)
+}
+
+.get_log10_table <- function(assay, target){
+
+    # If abundance table contains zeros, gives an error, because it is not possible
+    # to calculate log from zeros. If there is no zeros, calculates log.
+    if (any(assay == 0)) {
+        stop("Abundance table contains zero and log10 transformation
+                is being applied without pseudocount. Try log10 with
+                pseudocount=TRUE or log10p transformation.")
+    } else {
+        mat <- log10(assay)
+    }
+    return(mat)
+}
```

Comment:
What about `NA`s?

Reply:
NAs taken into account. Only case where NAs are replaced with 0 is when relative values are calculated. In that case, NAs are changed back to NAs after calculation. In other cases, NAs stay but there are na.rm=TRUE so they don't cause problems